File: iOS_Mac/QuartzPicture/objc/VKQuartzPicture.m
#import "VKQuartzPicture.h"

VKQuartzPicture VKQuartzPictureNull() {
    VKQuartzPicture picture = { CGRectNull, NULL };
    return picture;
}

void VKQuartzPictureDrawInRect(VKQuartzPicture picture, CGRect rect, CGContextRef context, VKPictureContentMode mode) {
    if (picture.drawer == NULL) {
        return;
    }

    if (picture.bounds.size.width < FLT_EPSILON || picture.bounds.size.height < FLT_EPSILON) {
        return;
    }

    CGFloat scaleX = rect.size.width / picture.bounds.size.width;
    CGFloat scaleY = rect.size.height / picture.bounds.size.height;

    switch (mode) {
        case VKPictureContentModeScaleAspectFit:
            scaleX = MIN(scaleX, scaleY);
            scaleY = scaleX;
            break;

        case VKPictureContentModeScaleAspectFill:
            scaleX = MAX(scaleX, scaleY);
            scaleY = scaleX;
            break;

        case VKPictureContentModeScaleToFill:
            // nothing
            break;

        case VKPictureContentModeCenter:
            scaleX = 1.0;
            scaleY = 1.0;
            break;
    }

    CGSize pictureSize = picture.bounds.size;
    pictureSize.width *= scaleX;
    pictureSize.height *= scaleY;

    CGPoint offset;
    offset.x = (rect.size.width - pictureSize.width) * 0.5 + rect.origin.x;
    offset.y = (rect.size.height - pictureSize.height) * 0.5 + rect.origin.y;

    CGContextSaveGState(context);
    CGContextTranslateCTM(context, offset.x, offset.y);
    CGContextScaleCTM(context, scaleX, scaleY);
    CGContextTranslateCTM(context, -picture.bounds.origin.x, -picture.bounds.origin.y);
    picture.drawer(context);
    CGContextRestoreGState(context);
}

File: iOS_Mac/QuartzPicture/swift/VKQuartzPicture.swift
import Foundation
import CoreGraphics

public enum VKPictureContentMode {
    case center
    case scaleToFill
    case scaleAspectFit
    case scaleAspectFill
}

public struct VKQuartzPicture {
    var bounds: CGRect
    var drawer: ((CGContext) -> Void)?
}

public func VKQuartzPictureNull() -> VKQuartzPicture {
    return VKQuartzPicture(bounds: CGRect.null, drawer: nil)
}

public func VKQuartzPictureDrawInRect(_ picture: VKQuartzPicture, _ rect: CGRect, _ context: CGContext, _ mode: VKPictureContentMode) {
    guard let drawer = picture.drawer else { return }

    if rect.size.width < CGFloat.ulpOfOne || rect.size.height < CGFloat.ulpOfOne {
        return
    }

    var scaleX: CGFloat = rect.size.width / picture.bounds.size.width
    var scaleY: CGFloat = rect.size.height / picture.bounds.size.height

    switch mode {
    case .scaleAspectFit:
        scaleX = min(scaleX, scaleY)
        scaleY = scaleX

    case .scaleAspectFill:
        scaleX = max(scaleX, scaleY)
        scaleY = scaleX

    case .scaleToFill:
        // nothing
        break

    case .center:
        scaleX = 1.0
        scaleY = 1.0
    }

    var pictureSize = picture.bounds.size
    pictureSize.width *= scaleX
    pictureSize.height *= scaleY

    var offset = CGPoint.zero
    offset.x = (rect.size.width - pictureSize.width) * 0.5 + rect.origin.x
    offset.y = (rect.size.height - pictureSize.height) * 0.5 + rect.origin.y

    context.saveGState()
    context.translateBy(x: offset.x, y: offset.y)
    context.scaleBy(x: scaleX, y: scaleY)
    context.translateBy(x: -picture.bounds.origin.x, y: -picture.bounds.origin.y)
    drawer(context)
    context.restoreGState()
}

